ribs/archives/_unstructured_archive.py

```diff
@@ -0,0 +1,524 @@
+            offset. This QD score offset will be *subtracted* from all
+            objectives in the archive, e.g., if your objectives go as low as
+            -300, pass in -300 so that each objective will be transformed as
+            ``objective - (-300)``.
+        seed (int): Value to seed the random number generator. Set to None to
+            avoid a fixed seed.
+        dtype (str or data-type or dict): Data type of the solutions,
+            objectives, and measures. We only support ``"f"`` / ``np.float32``
+            and ``"d"`` / ``np.float64``. Alternatively, this can be a dict
+            specifying separate dtypes, of the form ``{"solution": <dtype>,
+            "objective": <dtype>, "measures": <dtype>}``.
+        extra_fields (dict): Description of extra fields of data that is stored
+            next to elite data like solutions and objectives. The description is
+            a dict mapping from a field name (str) to a tuple of ``(shape,
+            dtype)``. For instance, ``{"foo": ((), np.float32), "bar": ((10,),
+            np.float32)}`` will create a "foo" field that contains scalar values
+            and a "bar" field that contains 10D values. Note that field names
+            must be valid Python identifiers, and names already used in the
+            archive are not allowed.
+    """
+
+    def __init__(self,
+                 *,
+                 solution_dim,
+                 measure_dim,
+                 k_neighbors,
+                 sparsity_threshold,
+                 learning_rate=None,
+                 threshold_min=-np.inf,
+                 qd_score_offset=0.0,
+                 seed=None,
+                 dtype=np.float64,
+                 extra_fields=None):
+
+        ArchiveBase.__init__(
+            self,
+            solution_dim=solution_dim,
+            cells=0,
+            measure_dim=measure_dim,
+            learning_rate=learning_rate,
+            threshold_min=threshold_min,
+            qd_score_offset=qd_score_offset,
+            seed=seed,
+            dtype=dtype,
+            extra_fields=extra_fields,
+        )
+
+        self._k_neighbors = int(k_neighbors)
+        self._sparsity_threshold = np_scalar(sparsity_threshold,
+                                             dtype=self.dtypes["measures"])
+
+    @property
+    def k_neighbors(self) -> int:
+        """int: The number of nearest neighbors to use for determining
+        sparsity."""
+        return self._k_neighbors
+
+    @property
+    def sparsity_threshold(self) -> float:
+        """:attr:`dtype` : The degree of sparseness in metric space required for
+        a solution to be added unconditionally."""
+        return self._sparsity_threshold
+
+    def index_of(self, measures, resize: bool = False) -> np.ndarray:
+        """Returns archive indices for the given batch of measures.
+
+        First, if the archive is empty it is resized to fit the new data.
```

Comment:
In general, batch insertion behavior for novelty search has not really been studied before; prior works assume you are only adding one solution at a time. For example, the Archive class in QDax just defines an iteration function and uses `jax.lax.scan` to apply it over the entire batch of new solutions https://github.com/adaptive-intelligent-robotics/QDax/blob/eb2cf317c7a5b6ffe385f0606a99e7c509b771b3/qdax/core/containers/archive.py#L134 Thus, we have some leeway in how we choose to handle things, but it should all be an extension of the original novelty search mechanisms, namely, if we insert one solution at a time with this method, we should get the same behavior as the original novelty search.

That being said, for the initial insertion, I don't think we should just insert all the solutions at once. That would mean we have solutions that fail to satisfy the novelty threshold and are too close together.

For subsequent insertions, we can find the k-nearest neighbors among both the archive and the incoming measures, and only pick solutions that satisfy the novelty threshold. I think this is what you are saying with computing distances to archive solutions and other incoming solutions and concatenating the distances. However, it can be put more succinctly by taking the whole set of `archive + incoming measures` together. When we look at it this way, the initial insertion is actually just the same as subsequent insertions, as the initial insertion will just have an incoming archive. That should help simplify this method.

Finally, I think you can make things a lot cleaner and faster by using scipy's cKDTree (or just KDTree nowadays), specifically the query method: https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.cKDTree.query.html#scipy.spatial.cKDTree.query